File: app/src/main/java/com/softartdev/notecrypt/ui/settings/security/SecurityPresenter.java
package com.softartdev.notecrypt.ui.settings.security;

import android.text.TextUtils;

import com.softartdev.notecrypt.db.DbStore;
import com.softartdev.notecrypt.di.ConfigPersistent;
import com.softartdev.notecrypt.ui.base.BasePresenter;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

@ConfigPersistent
class SecurityPresenter extends BasePresenter<SecurityView> {
    private DbStore dbStore;

    @Inject
    SecurityPresenter(DbStore dbStore) {
        this.dbStore = dbStore;
    }

    @Override
    public void attachView(@NotNull SecurityView mvpView) {
        super.attachView(mvpView);
    }

    boolean isEncryption() {
        return dbStore.isEncryption();
    }

    private boolean checkPass(String pass) {
        return dbStore.checkPass(pass);
    }

    private void changePassword(String odlPass, String newPass) {
        dbStore.changePass(odlPass, newPass);
    }

    void changeEncryption(boolean checked) {
        if (checked) {
            getMvpView().showSetPasswordDialog();
        } else {
            if (isEncryption()) {
                getMvpView().showPasswordDialog();
            } else {
                getMvpView().showEncryptEnable(false);
            }
        }
    }

    void changePassword() {
        if (isEncryption()) {
            getMvpView().showChangePasswordDialog();
        } else {
            getMvpView().showSetPasswordDialog();
        }
    }

    // only to disable encryption
    boolean enterPassCorrect(SecurityView.DialogDirector pass) {
        pass.hideError();

        String password = pass.getTextString();
        if (TextUtils.isEmpty(password)) {
            pass.showEmptyPasswordError();
            getMvpView().showEncryptEnable(true);
        } else if (checkPass(password)) {
            changePassword(password, null);
            getMvpView().showEncryptEnable(false);
            return true;
        } else {
            pass.showIncorrectPasswordError();
            getMvpView().showEncryptEnable(true);
        }

        return false;
    }

    // only to enable encryption
    boolean setPassCorrect(SecurityView.DialogDirector pass, SecurityView.DialogDirector repeatPass) {
        pass.hideError();
        repeatPass.hideError();

        String password = pass.getTextString();
        String repeatPassword = repeatPass.getTextString();
        if (password.equals(repeatPassword)) {
            if (TextUtils.isEmpty(password)) {
                pass.showEmptyPasswordError();
            } else {
                changePassword(null, password);
                getMvpView().showEncryptEnable(true);
                return true;
            }
        } else {
            repeatPass.showPasswordsNoMatchError();
        }

        return false;
    }

    // only when encryption is enabled
    boolean changePassCorrect(SecurityView.DialogDirector oldPass, SecurityView.DialogDirector newPass, SecurityView.DialogDirector repeatNewPass) {
        oldPass.hideError();
        newPass.hideError();
        repeatNewPass.hideError();

        String oldPassword = oldPass.getTextString();
        String newPassword = newPass.getTextString();
        String repeatNewPassword = repeatNewPass.getTextString();
        if (TextUtils.isEmpty(oldPassword)) {
            oldPass.showEmptyPasswordError();
        } else if (checkPass(oldPassword)) {
            if (TextUtils.isEmpty(newPassword)) {
                newPass.showEmptyPasswordError();
            } else if (newPassword.equals(repeatNewPassword)) {
                changePassword(oldPassword, newPassword);
                return true;
            } else {
                repeatNewPass.showPasswordsNoMatchError();
            }
        } else {
            oldPass.showIncorrectPasswordError();
        }

        return false;
    }
}

File: app/src/main/java/com/softartdev/notecrypt/ui/settings/security/SecurityPresenter.kt
package com.softartdev.notecrypt.ui.settings.security

import android.text.TextUtils
import com.softartdev.notecrypt.data.DataManager
import com.softartdev.notecrypt.di.ConfigPersistent
import com.softartdev.notecrypt.ui.base.BasePresenter
import timber.log.Timber
import javax.inject.Inject

@ConfigPersistent
internal class SecurityPresenter @Inject
constructor(private val dataManager: DataManager) : BasePresenter<SecurityView>() {

    override fun attachView(mvpView: SecurityView) {
        super.attachView(mvpView)
    }

    val isEncryption: Boolean
        get() = dataManager.isEncryption()

    private fun checkPass(pass: String?): Boolean {
        return dataManager.checkPass(pass)
    }

    private fun changePassword(odlPass: String?, newPass: String?) {
        addDisposable(dataManager.changePass(odlPass, newPass)
                .subscribe(
                        { Timber.d("Password changed") }
                        , { it.printStackTrace() }))
    }

    fun changeEncryption(checked: Boolean) {
        checkViewAttached()
        if (checked) {
            mvpView!!.showSetPasswordDialog()
        } else {
            if (isEncryption) {
                mvpView!!.showPasswordDialog()
            } else {
                mvpView!!.showEncryptEnable(false)
            }
        }
    }

    fun changePassword() {
        checkViewAttached()
        if (isEncryption) {
            mvpView!!.showChangePasswordDialog()
        } else {
            mvpView!!.showSetPasswordDialog()
        }
    }

    // only to disable encryption
    fun enterPassCorrect(pass: SecurityView.DialogDirector): Boolean {
        checkViewAttached()
        pass.hideError()

        val password = pass.textString
        if (TextUtils.isEmpty(password)) {
            pass.showEmptyPasswordError()
            mvpView!!.showEncryptEnable(true)
        } else if (checkPass(password)) {
            changePassword(password, null)
            mvpView!!.showEncryptEnable(false)
            return true
        } else {
            pass.showIncorrectPasswordError()
            mvpView!!.showEncryptEnable(true)
        }

        return false
    }

    // only to enable encryption
    fun setPassCorrect(pass: SecurityView.DialogDirector, repeatPass: SecurityView.DialogDirector): Boolean {
        checkViewAttached()
        pass.hideError()
        repeatPass.hideError()

        val password = pass.textString
        val repeatPassword = repeatPass.textString
        if (password == repeatPassword) {
            if (TextUtils.isEmpty(password)) {
                pass.showEmptyPasswordError()
            } else {
                changePassword(null, password)
                mvpView!!.showEncryptEnable(true)
                return true
            }
        } else {
            repeatPass.showPasswordsNoMatchError()
        }

        return false
    }

    // only when encryption is enabled
    fun changePassCorrect(oldPass: SecurityView.DialogDirector, newPass: SecurityView.DialogDirector, repeatNewPass: SecurityView.DialogDirector): Boolean {
        oldPass.hideError()
        newPass.hideError()
        repeatNewPass.hideError()

        val oldPassword = oldPass.textString
        val newPassword = newPass.textString
        val repeatNewPassword = repeatNewPass.textString
        if (TextUtils.isEmpty(oldPassword)) {
            oldPass.showEmptyPasswordError()
        } else if (checkPass(oldPassword)) {
            if (TextUtils.isEmpty(newPassword)) {
                newPass.showEmptyPasswordError()
            } else if (newPassword == repeatNewPassword) {
                changePassword(oldPassword, newPassword)
                return true
            } else {
                repeatNewPass.showPasswordsNoMatchError()
            }
        } else {
            oldPass.showIncorrectPasswordError()
        }

        return false
    }
}
